### PR TITLE
Cherry-pick 'Refactor virtual site handler' (#1252 and #1284)

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description

This PR cherry-picks the relevant parts of #1252 and #1284 into the biopolymer topology branch. The `create_force` changes and associated tests are omitted as this code will now live in `openff-interchange`.

Further, this PR drops all remaining v-site classes as v-sites will now be directly accessed on `Interchange` objects as they exist on a 'post-parameterisation' topology.

This PR is expected to be associated with one in the `openff-interchange` repo that moves the new `create_force` code there.

Remaining to-dos:
- [ ] Ensure that tests from `test_parameters` highlighted at [this link](https://github.com/openforcefield/openff-toolkit/pull/1252/files#diff-4b439486f53f6194c05e5906fb7a0e7006040444c4f97cee09f697822ec8340aR2445-R2771) get ported to Interchange
- [ ] Re-enable vsite_showcase notebook and get it to a working state

### CODE REVIEW

**Comment by mattwthompson** [comment]
Will #1253 be included as well?

**Comment by mattwthompson** [comment]
This would be pretty hard to review, and I'm not really sure where to start. I'd personally be happy giving this an "approval" on the basis that the tests are passing (failures I saw were unrelated) and you're using some form of this in fitting experiments now. The only exception might be `examples/virtual_sites/vsite_showcase.ipynb`, I don't know the plans there but I'd be okay with removing that, moving it to a deprecated path, etc.

I haven't updated Interchange to use this code yet, but I think it would be the wrong order of operations to use that development as a testing ground for this. I'd like to have this merged more or less as-is, and if I find changes are needed we can figure it out from there.

Does this sound like an okay plan to you (and @j-wags)?

**Review by j-wags — COMMENTED** [review]
Thanks for leading this @SimonBoothroyd and for engaging @mattwthompson. I could use more time to review this for unintended interactions with the topology refactor. I'm happy to be the only reviewer and take ownership of this PR.

File: openff/toolkit/tests/test_parameters.py [hunk 1]
```diff
@@ -2188,6 +2181,24 @@ def generate_v_site_coordinates(
...
+        )
+        handler.add_parameter(
+            parameter=VirtualSiteMocking.divalent_parameter("[*:2][*:1][*:3]", "once")
+        )
+        handler.add_parameter(
+            parameter=VirtualSiteMocking.trivalent_parameter("[*:1][*:2][*:3][*:4]")
+        )
+
     @pytest.mark.parametrize(
         "smiles, matched_indices, parameter, expected_raises",
```
Comment (j-wags) [inline openff/toolkit/tests/test_parameters.py:2114]: Note to self: I think this test is either misnamed or incomplete, I will fix later

File: openff/toolkit/tests/test_parameters.py [hunk 2]
```diff
@@ -2092,95 +2092,6 @@ class TestVirtualSiteHandler:
     Test the creation of a VirtualSiteHandler and the implemented VirtualSiteTypes
     """
 
-    @classmethod
```
Comment (j-wags) [inline openff/toolkit/tests/test_parameters.py]: Note to self: See if there's a better way to satisfy mypy other than the removals in https://github.com/openforcefield/openff-toolkit/pull/1285/commits/59598995dd889b284dea5e982f3d8a6d970fcd54

**Review by j-wags — COMMENTED** [review]
Jotting this down because it's not evident in the diffs - There are two super important tests in #1252 that aren't included here, and I think the intent was that they'd get moved over to interchange. They are linked here: https://github.com/openforcefield/openff-toolkit/pull/1252/files#diff-4b439486f53f6194c05e5906fb7a0e7006040444c4f97cee09f697822ec8340aR2445-R2771

**Comment by j-wags** [comment]
Ok, and resolving the failing example will also require some coordination with the interchange implementation. I'll work on this with @mattwthompson during our check in tomorrow.

**Review by j-wags — APPROVED** [review]
Alright - @mattwthompson and I just met to discuss how to move forward with this PR. We decided to merge immediately so that we avoid recursive development and have one less degree in the branch-of-a-branch-of-a-brach style development we're doing now. Matt will get vsites implemented in interchange based on current state of this branch.

There are two additional to-do items that need to be resolved before we make the 0.11.0 release, and I've added them as incomplete checklist items in the main text of this PR. Those will get inherited into the topology-biopolymer-refactor PR and marked as blocking so we make sure to resolve them before release.

**Pull request merged**